### PR TITLE
generate bonded interactions; this makes charmm topologies work

### PR DESCRIPTION
**Pull request by fgrunewald**
Some gromacs force-fields such as Charmm make use of bondtypes and angletypes etc. These need to be interpreted in the structure generator. The current pull request looks up a given parameter and replace them in the molecule.

### CODE REVIEW

File: polyply/src/topology.py [hunk 1]
```diff
@@ -216,14 +224,55 @@ def gen_pairs(self):
...
+                            new_params, meta = self.types[inter_type][("X", atoms[1], atoms[2], "X")]
+                        elif inter_type == "dihedrals" and\
+                             ("X", atoms[2], atoms[1], "X") in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][("X", atoms[2], atoms[1], "X")]
+                        elif inter_type in "dihedrals" and\
+                             (atoms[0], "X", "X", atoms[3]) in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][(atoms[0], "X", "X", atoms[3])]
+                        elif inter_type == "dihedrals" and\
+                             (atoms[3], "X", "X", atoms[0]) in self.types[inter_type]:
+                            new_params, meta = self.types[inter_type][(atoms[3], "X", "X", atoms[0])]
```
Comment (pckroon) [inline polyply/src/topology.py:266]: Can only dihedrals match X? And I'm sure there's a better way to organise this. I just don't know how (yet).

File: polyply/src/topology.py [hunk 2]
```diff
@@ -216,14 +224,55 @@ def gen_pairs(self):
...
+                        else:
+                            msg=("In section {} interaction of atoms {} has no corresponding bonded"
+                                 "type.")
+                            atoms = " ".join(list(map(lambda x: str(x), interaction.atoms)))
+                            raise OSError(msg.format(inter_type, atoms))
+
+                        interaction.parameters[:] = new_params[:]
```
Comment (pckroon) [inline polyply/src/topology.py:273]: See also https://docs.python.org/3.8/library/collections.html#collections.somenamedtuple._replace

File: polyply/tests/test_top_parser.py [hunk 3]
```diff
@@ -80,10 +80,9 @@ class TestTopParsing:
         [ bondtypes ]
         OM      O         1    1.9670816e-03  8.5679450e-07
         """,
-         "types",
-         defaultdict(list, {"bondtypes": [vermouth.molecule.Interaction(
-             atoms=["OM", "O"], parameters=["1", "1.9670816e-03", "8.5679450e-07"],
-             meta=None)]})
+        "types",
+        {"bonds": {("OM", "O"): (["1", "1.9670816e-03", "8.5679450e-07"],
+                         None)}}
```
Comment (pckroon) [inline polyply/tests/test_top_parser.py:85]: bad linebreak. I would prefer the line to be too long over this dangling None.

File: polyply/tests/test_top_parser.py [hunk 4]
```diff
@@ -161,6 +157,31 @@ def test_ifdefs(def_statements, bonds):
...
+        """
+        test files in ifdef/ifndef statements are only
+        read if a define has been read before
+        """
+        new_lines = textwrap.dedent(def_statements)
+        new_lines = new_lines.splitlines()
+        force_field = vermouth.forcefield.ForceField(name='test_ff')
+        top = Topology(force_field, name="test")
+        polyply.src.top_parser.read_topology(new_lines, top)
+        assert True
```
Comment (pckroon) [inline polyply/tests/test_top_parser.py:183]: Can't we do something better?

File: polyply/tests/test_top_parser.py [hunk 5]
```diff
@@ -272,3 +293,21 @@ def test_buckingham_fail():
...
+        [implicit_genborn_params]
+        something something
+        """
+         ))
+    def test_skip_directives(lines):
+        new_lines = textwrap.dedent(lines)
+        new_lines = new_lines.splitlines()
+        force_field = vermouth.forcefield.ForceField(name='test_ff')
+        top = Topology(force_field, name="test")
+        assert True
```
Comment (pckroon) [inline polyply/tests/test_top_parser.py:313]: ditto.
At least needs a comment that you're making sure there's no error, and why there shouldn't be.

File: polyply/tests/test_topology.py [hunk 6]
```diff
@@ -238,3 +239,206 @@ def test_convert_nonbond_to_sig_eps():
...
+        1.0   1.0   yes  1.0     1.0
+        [ bondtypes ]
+        C       C       1       0.1335  502080.0
+        [ moleculetype ]
+        test 3
+        [ atoms ]
+        1 C   1 test C1 1   0.0 14.0
+        2 C   1 test C2 2   0.0 12.0
+        [ bonds ]
+        1 2  1
```
Comment (pckroon) [inline polyply/tests/test_topology.py:256]: Add a case where a matching interaction already defines *other* parameters.

File: polyply/tests/test_topology.py [hunk 7]
```diff
@@ -238,3 +239,206 @@ def test_convert_nonbond_to_sig_eps():
...
+                                meta={})]}
+        ),
+        # test generic improper
+        (
+        """
+        [ defaults ]
+        1.0   1.0   yes  1.0     1.0
+        [ dihedraltypes ]
+        CT2   X     X     CT2    5   123.50	401.664	0.0	0.0
+        CT2   X     X     CE2    5   123.50	401.664	0.0	0.0
```
Comment (pckroon) [inline polyply/tests/test_topology.py:346]: Give htem different param values.
Add a non-matching pattern. Add a pattern that cannot match, e.g. X X X X ...
Add a matching pattern without X's. Which should be used? Which is used?

**Review by pckroon — CHANGES_REQUESTED** [review]
The tests can be much tighter, looks fine otherwise. I would've preferred the linting fixes in a separate PR though.
As an aside, do you need to do this matching, or can you leave it to Grompp?

File: polyply/src/topology.py (hunk 1 above)
Comment (fgrunewald) [reply polyply/src/topology.py:266]: yes only dihedrals you can't write angles or bonds like this. I tested it and grompp fails.

File: polyply/tests/test_top_parser.py (hunk 4 above)
Comment (fgrunewald) [reply polyply/tests/test_top_parser.py:183]: In what sense better? All it is supposed to do is do nothing meaning it doesn't fail. I can't imagine really a more through test for this.

File: polyply/tests/test_topology.py (hunk 7 above)
Comment (fgrunewald) [reply polyply/tests/test_topology.py:346]: >  Add a matching pattern without X's. Which should be used? Which is used?< 

GROMACS goes by priority or position in the itp file. There is no documentation for this.

File: polyply/tests/test_top_parser.py (hunk 5 above)
Comment (fgrunewald) [reply polyply/tests/test_top_parser.py:313]: I added a doc-string. I think for now these tests test what they should. Also this is very obscure grompp behavior which I think is purely for legacy reasons.

File: polyply/tests/test_topology.py (hunk 7 above)
Comment (fgrunewald) [reply polyply/tests/test_topology.py:346]: so gromacs takes the more specific interaction if defined.

File: polyply/src/topology.py (hunk 2 above)
Comment (fgrunewald) [reply polyply/src/topology.py:273]: this returns a new instance but I want to keep the old one and just overwrite the parameters

Comment (pckroon) [reply polyply/src/topology.py:273]: This on the other hand will only work when the parameters are lists, which is not defined. See also https://github.com/marrink-lab/vermouth-martinize/issues/240

File: polyply/tests/test_top_parser.py (hunk 4 above)
Comment (pckroon) [reply polyply/tests/test_top_parser.py:182]: Make sure random.itp exists, and that the things it contains do not end up in the FF. On the other hand, maybe it's good that random.itp doesn't exist. That way you can be sure parsing isn't even tried.

Better to leave it as is.

File: polyply/tests/test_topology.py [hunk 8]
```diff
@@ -333,7 +334,88 @@ def test_convert_nonbond_to_sig_eps():
         """,
         {"dihedrals": [Interaction(atoms=(0, 1, 2, 3), parameters=["5", "123.50", "401.664", "0.0", "0.0"],
                                 meta={}),
-                       Interaction(atoms=(1, 2, 3, 4), parameters=["5", "123.50", "401.664", "0.0", "0.0"],
+                       Interaction(atoms=(1, 2, 3, 4), parameters=["5", "120", "400", "0.0", "0.0"],
+                                meta={})]}
```
Comment (pckroon) [inline polyply/tests/test_topology.py:338]: bad linebreak

File: polyply/tests/test_topology.py [hunk 9]
```diff
@@ -333,7 +334,88 @@ def test_convert_nonbond_to_sig_eps():
...
+        2  3  4  5 1   150  60  0.0 0.0
+        [ system ]
+        some title
+        [ molecules ]
+        test 1
+        """,
+        {"dihedrals": [Interaction(atoms=(0, 1, 2, 3), parameters=["5", "123.50", "401.664", "0.0", "0.0"],
+                                meta={}),
+                       Interaction(atoms=(1, 2, 3, 4), parameters=["1", "150", "60", "0.0", "0.0"],
+                                meta={})]}
```
Comment (pckroon) [inline polyply/tests/test_topology.py:367]: I also don't like these linebreaks

File: polyply/tests/test_topology.py [hunk 10]
```diff
@@ -333,7 +334,88 @@ def test_convert_nonbond_to_sig_eps():
...
+        5 CT2   1 test C5 5   0.0 14.0
+        [ dihedrals ]
+        1  2  3  4 1
+        [ system ]
+        some title
+        [ molecules ]
+        test 1
+        """,
+        {"dihedrals": [Interaction(atoms=(0, 1, 2, 3), parameters=["5", "123.50", "401.664", "0.0", "0.0"],
+                                meta={})]}
```
Comment (pckroon) [inline polyply/tests/test_topology.py:393]: ditto

File: polyply/tests/test_topology.py [hunk 11]
```diff
@@ -333,7 +334,88 @@ def test_convert_nonbond_to_sig_eps():
...
+        4 CT2   1 test C4 4   0.0 12.0
+        5 CT2   1 test C5 5   0.0 14.0
+        [ dihedrals ]
+        1  2  3  4 1
+        [ system ]
+        some title
+        [ molecules ]
+        test 1
+        """,
+        {"dihedrals": [Interaction(atoms=(0, 1, 2, 3), parameters=["5", "123.50", "401.664", "0.0", "0.0"],
```
Comment (pckroon) [inline polyply/tests/test_topology.py:418]: idem

**Review by pckroon — APPROVED** [review]
Small linting nitpicks. LGTM otherwise :+1:

File: polyply/src/top_parser.py [hunk 12]
```diff
@@ -308,10 +308,33 @@ def _atomtypes(self, line, lineno=0):
...
+             if var != "ptype":
+                atom_type_line[var] = float(tokens.pop(-1))
+             else:
+                atom_type_line[var] = tokens.pop(-1)
+
+        if len(tokens) >= 1:
+            atom_type_line["atom_num"] = float(tokens.pop(-1))
+
+        if len(tokens) == 1:
+            atom_type_line["bond_type"] = tokens.pop(0)
```
Comment (pckroon) [inline polyply/src/top_parser.py:331]:
```suggestion
        from itertools import zip_longest
        atom_type_line = dict(zip_longest(["atom_name", "nb2", "nb1", "ptype", "charge", "mass"], tokens)
        if None in atom_type_line:
            msg = ("Can't parse line {}. Found more parameters than expected.")
            raise OSError(msg.format(line))
```
Something along those lines anyway. Maybe zip with a defaultdict is better than zip_longest.

File: polyply/src/topology.py [hunk 13]
```diff
@@ -246,7 +246,12 @@ def gen_bonded_interactions(self):
                     continue
                 for interaction in interactions:
                     if len(interaction.parameters) == 1:
-                        atoms = tuple(block.nodes[node]["atype"] for node in interaction.atoms)
+
+                        if "_FF_OPLS" in self.defines or "_FF_OPLS_AA" in self.defines:
+                           atoms = tuple(self.atom_types[block.nodes[node]["atype"]]["bond_type"]
+                                         for node in interaction.atoms)
+                        else:
+                           atoms = tuple(block.nodes[node]["atype"] for node in interaction.atoms)
```
Comment (pckroon) [inline polyply/src/topology.py:254]: indentlint

Comment (pckroon) [reply polyply/src/topology.py:254]: Needs comment, I can't unpack this.

**Review by pckroon — CHANGES_REQUESTED** [review]

File: polyply/src/top_parser.py (hunk 12 above)
Comment (fgrunewald) [reply polyply/src/top_parser.py]: I think I captured the essence of what you suggested.

**Review by pckroon — APPROVED** [review]

**Pull request merged**